add sorting option for saved connections

caohai opened the pull request:
<!-- Thank you for submitting a Pull Request. Please:
* Read our Pull Request guidelines:
  https://github.com/Microsoft/azuredatastudio/wiki/How-to-Contribute#pull-requests.
* Associate an issue with the Pull Request.
* Ensure that the code is up-to-date with the `main` branch.
* Include a description of the proposed changes and how to test them.
-->

This PR fixes #2799 by adding a user setting (enum) `datasource.connections.sortBy` for connection sorting.

This setting is default to `dateAdded` which aligns with the current implementation so the default order for connections in Saved Connections tab remains unchanged.

Users will be able to select `displayName` which sort saved connections and groups by their display names.

Before:
![demo_15229_before_compressed](https://user-images.githubusercontent.com/21186993/116169796-609d9500-a6ba-11eb-8415-997c97cee47a.gif)

After:
![demo_15229_after_compressed](https://user-images.githubusercontent.com/21186993/116169972-b96d2d80-a6ba-11eb-8632-01819a572e6a.gif)

Charles-Gagnon commented on src/sql/workbench/contrib/dataExplorer/browser/dataExplorer.contribution.ts:
```diff
@@ -42,6 +42,12 @@ configurationRegistry.registerConfiguration({
+		},
+		'datasource.connectionSort': {
+			'type': 'string',
+			'enum': ['by time added', 'by title alphabetically'],
```
The convention we follow is that the enum values are camel cased with no spaces (e.g. `byTimeAdded` and `byTitleAlphabetically`. You can do a search for `'enum':` to see other examples of this in the code base

caohai replied:
updated

Charles-Gagnon commented on src/sql/workbench/contrib/dataExplorer/browser/dataExplorer.contribution.ts:
```diff
@@ -42,6 +42,12 @@ configurationRegistry.registerConfiguration({
+		'datasource.connectionSort': {
+			'type': 'string',
+			'enum': ['by time added', 'by title alphabetically'],
+			'default': 'by time added',
```
Typically we should always include an `enumDescriptions` field too which contains the localized descriptions of each of the enum values. See https://github.com/microsoft/azuredatastudio/blob/main/src/sql/workbench/contrib/notebook/browser/notebook.contribution.ts for an example

caohai replied:
added

Charles-Gagnon commented on src/sql/platform/connection/common/connectionConfig.ts:
```diff
@@ -214,6 +215,25 @@ export class ConnectionConfig {
+			});
+		}
+
+		if (sortFunc) {
```
Could just call this directly in the if check if the sortBy setting is by title

caohai replied:
I put it this way mainly for potential new sort options other than by title so we can add checks for those in this way. But if `by title` is the only sort option in foreseeable future then sure we can call it directly.

Charles-Gagnon replied:
Yeah that's fair - it's not a big deal so if you want to leave it like this that's fine. I generally prefer going for the simplest/easiest to read option unless we have actual plans to add others (which I don't believe we have) but go with whatever makes the most sense to you.

caohai replied:
I'll leave it as it is to allow for other potential options.

Charles-Gagnon reviewed (commented):
Your screenshots show the updated saved connections list but what about the server treeview in the connections viewlet? That seems to be the primary place that people are asking to have this list be sortable. 

(and if this is reflected there - does the view get updated immediately when the user changes the setting? Usually we'd have to hook into the configuration change event to ensure that is updated without requiring a restart)

This function is also called from a lot of different places so it would be good to do a quick check of those to make sure that we're ok with the results possibly being ordered alphabetically there as well. And this is something that we should consider scoping to just the places we specifically want to display it alphabetically to reduce the overall impact (I don't see this breaking any scenarios but you never know...)

caohai commented:
> Your screenshots show the updated saved connections list but what about the server treeview in the connections viewlet? That seems to be the primary place that people are asking to have this list be sortable.
> 
> (and if this is reflected there - does the view get updated immediately when the user changes the setting? Usually we'd have to hook into the configuration change event to ensure that is updated without requiring a restart)
> 
> This function is also called from a lot of different places so it would be good to do a quick check of those to make sure that we're ok with the results possibly being ordered alphabetically there as well. And this is something that we should consider scoping to just the places we specifically want to display it alphabetically to reduce the overall impact (I don't see this breaking any scenarios but you never know...)



> Your screenshots show the updated saved connections list but what about the server treeview in the connections viewlet? That seems to be the primary place that people are asking to have this list be sortable.
> 
> (and if this is reflected there - does the view get updated immediately when the user changes the setting? Usually we'd have to hook into the configuration change event to ensure that is updated without requiring a restart)
> 
> This function is also called from a lot of different places so it would be good to do a quick check of those to make sure that we're ok with the results possibly being ordered alphabetically there as well. And this is something that we should consider scoping to just the places we specifically want to display it alphabetically to reduce the overall impact (I don't see this breaking any scenarios but you never know...)

Regarding the left-side server treeview in the connections viewlet, it's updated after a reload, no immediately. I'll take a look at configuration change event to address this.

This function is essentially called by `getConnectionProfileGroups()` in `ConnectionStore` when building connection groups. I'll take another look to make sure this change won't break other places.

alanrenmsft commented:
the saved connections view will be removed and merged into the Browse tab, please make sure it is also ordered there

caohai commented:
> the saved connections view will be removed and merged into the Browse tab, please make sure it is also ordered there

With current implementation, the result in Browser tab is sorted if this setting is updated before the connection dialog is loaded. After that a window reload is needed for the new sort setting to take effective. 

I'll take a look at configuration change event to ensure the list is updated immediately without a reload.

caohai commented:
@Charles-Gagnon Hi Charles, I've addressed comments and added configuration change event to update OE and Browse tab immediately without a reload. Demo gifs are updated as well.

As for other places calling this `getConnections()` function, I've looked at the OE server view and Connection dialog, connection lists look good to me. Do you think I'm missing any other places?

@alanrenmsft Hi Alan, I've checked Browser tab and it's immediately sorted after `byTitleAlphabetically` is toggled and the results are consistent across OE, Saved Connections tab and Saved Connections node in Browse tab. You can check the demo gif.

@kburtram Hi Karl, I've added group sorting and its test case. Can you take a look (including other changes)?

alanrenmsft commented on src/sql/workbench/contrib/dataExplorer/browser/dataExplorer.contribution.ts:
```diff
@@ -42,6 +42,16 @@ configurationRegistry.registerConfiguration({
 			'description': localize('datasource.connectionGroups', "data source groups"),
 			'type': 'array'
+		},
+		'datasource.connectionSort': {
```
datasource.connections.sortby?

caohai replied:
updated

alanrenmsft commented on src/sql/workbench/contrib/dataExplorer/browser/dataExplorer.contribution.ts:
```diff
@@ -42,6 +42,16 @@ configurationRegistry.registerConfiguration({
+		},
+		'datasource.connectionSort': {
+			'type': 'string',
+			'enum': ['byTimeAdded', 'byTitleAlphabetically'],
```
'DateAdded', 'Name'

Charles-Gagnon replied:
And specifically I think he's suggesting `byDateAdded` and `byName` for the enum names. 

And maybe even `byDisplayName` to clearly indicate that it's the name shown in the OE tree - which is not necessarily the server name. 

(title is unclear what that means since we don't refer to title in the UI at all)

caohai replied:
updated
@Charles-Gagnon With the setting key updated to `sortBy` from `connectionSort`, I used `DateAdded` and `DisplayName` as enum names.

alanrenmsft commented on src/sql/workbench/contrib/dataExplorer/browser/dataExplorer.contribution.ts:
```diff
@@ -42,6 +42,16 @@ configurationRegistry.registerConfiguration({
 			'description': localize('datasource.connectionGroups', "data source groups"),
 			'type': 'array'
+		},
+		'datasource.connectionSort': {
```
I know it is not started by you, but why not use the const string defined above?

caohai replied:
updated, fixed other const string keys as well.

Charles-Gagnon commented on src/sql/workbench/contrib/objectExplorer/browser/serverTreeView.ts:
```diff
@@ -196,6 +196,11 @@ export class ServerTreeView extends Disposable implements IServerTreeView {
 			}
 		}));
+		this._register(this._configurationService.onDidChangeConfiguration(e => {
+			if (e.affectsConfiguration('datasource.connectionSort')) {
```
Should use `CONNECTION_SORT_CONFIG_KEY` here

caohai replied:
updated

Charles-Gagnon commented on src/sql/workbench/services/connection/browser/connectionBrowseTab.ts:
```diff
@@ -232,6 +233,13 @@ export class ConnectionBrowserView extends Disposable implements IPanelView {
 		}));
+
+		this._register(this.configurationService.onDidChangeConfiguration(e => {
+			if (e.affectsConfiguration('datasource.connectionSort')) {
```
`CONNECTION_SORT_CONFIG_KEY`

caohai replied:
updated

Charles-Gagnon commented on src/sql/platform/connection/test/common/connectionConfig.test.ts:
```diff
@@ -384,6 +398,18 @@ suite('ConnectionConfig', () => {
 		});
 	});
 
+	test('getConnections should return connections sorted alphabetically by title given datasource.connectionSort is set to \'byTitleAlphabetically\'', () => {
```
Do we already have a test for the sorting by date?

caohai replied:
With current implementation, connections are by default sorted by date, so the existing test case should be sufficient for that.

caohai replied:
added test cases for sorting by date

Charles-Gagnon commented on src/sql/workbench/contrib/dataExplorer/browser/dataExplorer.contribution.ts:
```diff
@@ -42,6 +42,16 @@ configurationRegistry.registerConfiguration({
+			'type': 'string',
+			'enum': ['byTimeAdded', 'byTitleAlphabetically'],
+			'enumDescriptions': [
+				localize('connectionSort.byTimeAdded', 'Saved connections are sorted by the time they were added.'),
```
`by the date they were added`

caohai replied:
updated

Charles-Gagnon commented on src/sql/workbench/contrib/dataExplorer/browser/dataExplorer.contribution.ts:
```diff
@@ -42,6 +42,16 @@ configurationRegistry.registerConfiguration({
+			'enum': ['byTimeAdded', 'byTitleAlphabetically'],
+			'enumDescriptions': [
+				localize('connectionSort.byTimeAdded', 'Saved connections are sorted by the time they were added.'),
+				localize('connectionSort.byTitleAlphabetically', 'Saved connections are sorted by their titles alphabetically.')
```
This should be updated to correspond to the new enum value as well

caohai replied:
updated

Charles-Gagnon commented on src/sql/workbench/contrib/dataExplorer/browser/dataExplorer.contribution.ts:
```diff
@@ -35,13 +36,23 @@ configurationRegistry.registerConfiguration({
+		},
+		[CONNECTION_SORT_BY_CONFIG_KEY]: {
+			'type': 'string',
+			'enum': ['DateAdded', 'DisplayName'],
```
I would suggest defining these enum values in the same place as the config keys

```
export const enum ConnectionSortBy {
    dateAdded = 'dateAdded',
    displayName = 'displayName'
}
```

So that you can reference that enum instead of hardcoding the values in the checks above

(and note that the enum values by convention should be camelCased)

caohai replied:
updated, I also personally prefer this kind of definition over the repeated string literals so I don't need to memorize every place I used same config....

Charles-Gagnon commented on src/sql/workbench/contrib/dataExplorer/browser/dataExplorer.contribution.ts:
```diff
@@ -35,13 +36,23 @@ configurationRegistry.registerConfiguration({
+				localize('sortBy.DateAdded', 'Saved connections are sorted by the date they were added.'),
+				localize('sortBy.DisplayName', 'Saved connections are sorted by their display name alphabetically.')
+			],
+			'default': 'byTimeAdded',
```
You didn't update the default here

caohai replied:
updated

Charles-Gagnon commented on src/sql/workbench/contrib/dataExplorer/browser/dataExplorer.contribution.ts:
```diff
@@ -35,13 +36,23 @@ configurationRegistry.registerConfiguration({
+			'type': 'string',
+			'enum': ['DateAdded', 'DisplayName'],
+			'enumDescriptions': [
+				localize('sortBy.DateAdded', 'Saved connections are sorted by the date they were added.'),
```
`connections.sortBy.dateAdded` 
`connections.sortBy.displayName`

caohai replied:
updated. I also tried to eliminate string literals but `'connections.sortBy.' + ConnectionsSortBy.dateAdded` doesn't look any better than `connections.sortBy.dateAdded` so I just went with string literals here.

Charles-Gagnon replied:
Yeah string literals for the loc IDs are fine since those don't have to match anything specifically. It's more about making it so the ID is descriptive enough for what the string contains to avoid confusion.

Charles-Gagnon reviewed (changes requested):
Make sure you merge in from main, you have a merge conflict now with the latest changes made there

Charles-Gagnon commented on src/sql/workbench/contrib/dataExplorer/browser/dataExplorer.contribution.ts:
```diff
@@ -22,13 +23,23 @@ configurationRegistry.registerConfiguration({
+				localize('connections.sortBy.displayName', 'Saved connections are sorted by their display names alphabetically.')
+			],
+			'default': ConnectionsSortBy.dateAdded,
+			'description': localize(CONNECTIONS_SORT_BY_CONFIG_KEY, "Order used for sorting saved connections and connection groups")
```
Let's just use a string constant for the ID - that doesn't need to actually match the config key name and so we don't want to tie them together since they aren't really related at all (the config key could change but we don't want to have that result in a new "string" needing to be localized)

caohai replied:
updated

Charles-Gagnon reviewed (approved):
One comment about loc IDs that should be addressed, but otherwise looks good as a v1. 

But, as a follow up improvement what would probably be useful is to add a command to control the sort order so people can easily switch back and forth without needing to open up the settings pane. In addition to being in the command palette we could add it to the list of actions at the top of the server pane : 

![image](https://user-images.githubusercontent.com/28519865/116326193-b20c5980-a778-11eb-9ee8-5dd732bd8b13.png)

@radentonev Do you have any mockups for this kind of functionality?

kburtram reviewed (approved):
LGTM
